File: app/schemas/agent.py

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field, model_validator


class AgentImageInput(BaseModel):
    image_url: Optional[str] = None
    image_base64: Optional[str] = None
    mime_type: str = "image/png"

    @model_validator(mode="after")
    def validate_source(self) -> "AgentImageInput":
        if not self.image_url and not self.image_base64:
            raise ValueError("image_url or image_base64 is required")
        return self


class AgentIdentityVerification(BaseModel):
    phone: Optional[str] = None
    id_number: Optional[str] = None

    @model_validator(mode="after")
    def validate_credential(self) -> "AgentIdentityVerification":
        if not self.phone and not self.id_number:
            raise ValueError("phone or id_number is required")
        return self


class AgentAccessContext(BaseModel):
    actor_role: str
    access_purpose: str
    operator_id: str = "demo-operator"
    tenant_id: str = "demo-tenant"
    is_demo_context: bool = False


class AgentQueryRequest(BaseModel):
    model_config = ConfigDict(extra="ignore")

    query: str
    patient_code: Optional[str] = None
    identity_verification: Optional[AgentIdentityVerification] = None
    images: List[AgentImageInput] = Field(default_factory=list)
    debug_planner: bool = False
    force_full_agent: bool = False


class AgentToolOutput(BaseModel):
    tool_name: str
    arguments: Dict[str, Any]
    result: Dict[str, Any]


class AgentQueryResponse(BaseModel):
    answer: str
    tool_outputs: List[AgentToolOutput]
    intent: Optional[str] = None
    intent_confidence: Optional[float] = None
    route_type: Optional[str] = None
    route_reason: Optional[str] = None
    cache_hit: Optional[bool] = None
    evidence: List[Dict[str, Any]] = Field(default_factory=list)
    risk_level: Optional[str] = None
    recommended_action: Optional[str] = None
    runtime_metrics: Optional[Dict[str, Any]] = None
    planner_debug: Optional[Dict[str, Any]] = None
    run_id: Optional[str] = None
    memory_refresh_scheduled: Optional[bool] = None
    execution_trace: Optional[Dict[str, Any]] = None

File: frontend/src/types/agent.ts
export type AgentImageInput = {
  image_base64: string;
  mime_type: string;
};

export type AgentQueryRequest = {
  query: string;
  patient_code?: string | null;
  identity_verification?: {
    phone?: string;
    id_number?: string;
  } | null;
  images: AgentImageInput[];
  debug_planner: boolean;
  force_full_agent?: boolean;
};

export type AgentToolOutput = {
  tool_name: string;
  arguments: Record<string, unknown>;
  result: Record<string, unknown>;
};

export type AgentQueryResponse = {
  answer: string;
  tool_outputs: AgentToolOutput[];
  intent?: string | null;
  intent_confidence?: number | null;
  route_type?: string | null;
  route_reason?: string | null;
  cache_hit?: boolean | null;
  evidence?: Record<string, unknown>[];
  risk_level?: string | null;
  recommended_action?: string | null;
  runtime_metrics?: Record<string, unknown> | null;
  planner_debug?: Record<string, unknown> | null;
  run_id?: string | null;
  memory_refresh_scheduled?: boolean | null;
  execution_trace?: Record<string, unknown> | null;
};

export type ChatMessage = {
  id: string;
  role: "user" | "assistant";
  content: string;
  imageUrl?: string;
  toolOutputs?: AgentToolOutput[];
  intent?: string | null;
  intentConfidence?: number | null;
  routeType?: string | null;
  routeReason?: string | null;
  cacheHit?: boolean | null;
  evidence?: Record<string, unknown>[];
  riskLevel?: string | null;
  recommendedAction?: string | null;
  status?: "pending" | "error" | "done";
};
